apps/chat/bots.py

```diff
@@ -163,6 +165,42 @@ def _save_message_to_history(self, message: str, type_: ChatMessageType):
                 content=message,
             )
 
+    def _get_optimized_history(self):
+        try:
+            return compress_chat_history(self.chat, self.llm, self.max_token_limit)
+        except (NameError, ImportError, ValueError):
+            # typically this is because a library required to count tokens isn't installed
+            log.exception("Unable to compress history")
+            return self.chat.get_langchain_messages_until_summary()
+
+
+def compress_chat_history(chat: Chat, llm: BaseLanguageModel, max_token_limit: int, keep_history_len: int = 10):
+    """Compresses the chat history to be less than max_token_limit tokens long. This will summarize the history
+    if necessary and save the summary to the DB.
+    """
+    history = chat.get_langchain_messages_until_summary()
+    if max_token_limit <= 0 or not history:
```

Comment:
Should we not rather prevent the user from entering a token limit smaller than 0?

Reply:
I guess I could make the model field a PositiveSmallIntegerField but it's not really a big deal. The `<= 0` was just me being defensive.

Reply:
883d60f

Reply:
:+1: